ext/crates/fp/src/limb.rs

```diff
@@ -11,6 +10,148 @@ pub(crate) struct LimbBitIndexPair {
     pub(crate) bit_index: usize,
 }
 
+/// A struct that defines a range of entries in a slice of limbs. We number limbs starting from some
+/// other limb that might not be in the range. This is because we must be able to index a struct
+/// using indices for another ambient struct. For example, accessing a `Slice` using coordinates
+/// from an ambient `FpVector`.
+#[derive(Debug, Hash, PartialEq, Eq, Clone, Copy)]
+pub struct LimbLength<const P: u32> {
+    /// The index of the first entry relative to some other base limb.
+    pub(crate) start: usize,
+
+    /// The index of the last entry relative to some other base limb.
+    pub(crate) end: usize,
+
+    /// The total number of limbs in the range.
+    ///
+    /// We store this value instead of computing it on the fly because benchmarks tend to show that
+    /// the tradeoff is beneficial in high dimensions (>1000). We might want to only enable this
+    /// when odd-primes is enabled, since computing this number is easier when `p == 2`, so the
+    /// tradeoff is potentially worse.
+    limbs: usize,
+}
+
+impl<const P: u32> LimbLength<P> {
```

Comment:
It seems like a lot of the functions can be made `const`, e.g. `bit_offset`. It's probably good to make them so.

Reply:
I made a few more functions `const`. The others can't be because either they end up calling `pack` which has a for loop, or they call some non-const function in stdlib